src/hoa-convolver.js

```diff
@@ -0,0 +1,156 @@
+/**
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileOverview A collection of convolvers. Can be used for the optimized HOA
+ *               binaural rendering. (e.g. SH-MaxRe HRTFs)
+ */
+
+'use strict';
+
+/**
+ * @class HOAConvolver
+ * @description A collection of convolvers for N-channel HOA stream.
+ * @param {AudioContext} context          Associated AudioContext.
+ * @param {Object} options                Options for speaker.
+ * @param {Array} options.IR              Multichannel HRTF convolution buffer.
+ * @param {Number} options.ambisonicOrder Ambisonic order (default is 3).
+ * @param {Number} options.gain           Post-gain for the speaker (optional).
+ */
+function HOAConvolver (context, options) {
+  this._active = false;
+
+  this._context = context;
+
+  /**
+   * We need to determine the number of channels K based on the ambisonic
+   * order N where
+   *   K = (N + 1)^2
+   */
+  var ambisonicOrder = 3;
+  if (options.ambisonicOrder) {
+    ambisonicOrder = options.ambisonicOrder;
+  }
+  var numChannels = (ambisonicOrder + 1) * (ambisonicOrder + 1);
+
+  // Ensure that the ambisonic order matches the IR channel count.
+  if (options.IR.numberOfChannels !== numChannels) {
+    throw 'Ambisonic order and IR channel count do not match. cannot proceed.';
+  }
+
+  // Compute the number of stereo convolvers needed.
+  var numStereoChannels = Math.round(numChannels / 2);
+
+  this._input = this._context.createChannelSplitter(numChannels);
+  this._mergers = [];
+  this._convolvers = [];
+  this._splitters = [];
+  for (var i = 0; i < numStereoChannels; i++) {
+    this._mergers[i] = this._context.createChannelMerger(2);
```

Comment:
I need to split channels from 16-channel input into stereo chunks, so I'm taking every 2 channels from input and mergining them into a stereo stream, then running that through a convolverNode then splitting the stereo stream and placing the two streams back into a 16-channel output. Should I be going about this a better way? I'm trying to leverage the stereo convolver like you mentioned.